File: backend/app/core/upgrades/stat_modifiers/bullet_damage_5.py
from ..upgrade_implementation import UpgradeImplementation, UpgradeAppliers


class BulletDamage5(UpgradeImplementation):
    id = 'bullet_damage_5'
    name = 'Sharper Rounds'
    description = '+35 bullet damage'
    rarity = 'legendary'
    type = 'stat_modifier'
    cost = 40

    def can_apply(self, state: dict) -> bool:
        return True

    def apply(self, appliers: UpgradeAppliers) -> None:
        value = 35
        appliers.modifiers.add_multiplier('attack', 'damage', value)

    def validate(self, recorded_stats: dict) -> bool:
        return True

File: frontend/src/game/upgrades/stat_modifiers/bullet_damage_5.ts
import { Upgrade, type UpgradeDef } from '../Upgrade'
import { UpgradeTargetID, UpgradeStatID, RarityID, UpgradeTypeID } from '../../data/ID'

export const BulletDamage5Def: UpgradeDef = {
  id: "bullet_damage_5",
  name: "Sharper Rounds",
  description: "+35 bullet damage",
  rarity: RarityID.Legendary,
  upgradeType: UpgradeTypeID.StatModifier,
  cost: 40,
  targetClass: UpgradeTargetID.Bullet,
  fieldInTargetClass: UpgradeStatID.Damage,
  value: 35,
  isMultiplier: false,
  specificAttackType: "bullet",
  stackable: true,
  maxStacks: 99999,
}

export class BulletDamage5 extends Upgrade {}
